### ISSUE TITLE
test_ttl_is_respected_on_repair[parallel_aggregates_enabled] fails

### ISSUE THREAD

**Issue by tchaikov** [scylladb/scylladb#13276]
spotted in a promotion run against an unrelated PR:

```
ttl_test.py:787: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
tools/retrying.py:43: in inner
    return func(*args, **kwargs)
tools/assertions.py:254: in assert_row_count
    count = run_query_with_data_processing(session, query, consistency_level=consistency_level, session_timeout=timeout)
tools/data.py:334: in run_query_with_data_processing
    result = list(session.execute(SimpleStatement(query, consistency_level=consistency_level), timeout=session_timeout))
cassandra/cluster.py:2611: in cassandra.cluster.Session.execute
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.Unavailable: Error from server: code=1000 [Unavailable exception] message="Cannot achieve consistency level for cl ONE. Requires 1, alive 0" info={'consistency': 'ONE', 'required_replicas': 1, 'alive_replicas': 0}
```

see https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/188/testReport/junit/ttl_test/TestDistributedTTL/Sanity_Tests___test_ttl_is_respected_on_repair_parallel_aggregates_enabled_/

**Comment by mykaul** [comment]
I *think* it's a dup of https://github.com/scylladb/scylladb/issues/13137

**Comment by tchaikov** [comment]
nah, they are different. the error in #13137 looks like
```
got [[1, 1, None, None]]
```

while this one is something different.

**Comment by Jadw1** [comment]
Cc @bhalevy @eliransin

**Comment by bhalevy** [comment]
I haven't seen this after https://github.com/scylladb/scylla-dtest/pull/3086
I think it can be closed for now.

That said, there's a similar yet different issue I've seen in https://github.com/scylladb/scylla-dtest/blob/4b1e5c7369b5cd4742ea75c995fc2ae0627a454d/ttl_test.py#L787
bit it deserves a new github issue

**Comment by bhalevy** [comment]
https://jenkins.scylladb.com/job/scylla-enterprise/job/next/2320/testReport/junit/ttl_test/TestDistributedTTL/Tests___Sanity_Tests___test_ttl_is_respected_on_repair_parallel_aggregates_enabled_/
```
self = <ttl_test.TestDistributedTTL object at 0x7f18249cc8d0>
enable_parallized_aggregation = 'true'

    @pytest.mark.next_gating
    @pytest.mark.dtest_debug
    @pytest.mark.parametrize("enable_parallized_aggregation", ['true', 'false'], ids=["parallel_aggregates_enabled", "parallel_aggregates_disabled"])
    def test_ttl_is_respected_on_repair(self, enable_parallized_aggregation):
        """ Test that ttl is respected on repair """
    
        self.prepare(jvm_args=['--logger-log-level', 'forward_service=trace',
                     '--enable-parallelized-aggregation', enable_parallized_aggregation])
        nodes = self.cluster.nodelist()
        short_ttl = 10
        long_ttl = 1000
        change_schema_safely(self.session1, nodes, """
            ALTER KEYSPACE ks WITH REPLICATION =
            {'class' : 'SimpleStrategy', 'replication_factor' : 1};
        """)
    
        self.session1.execute(f"""
            INSERT INTO ttl_table (key, col1) VALUES (1, 1) USING TTL {short_ttl};
        """)
        self.session1.execute(f"""
            INSERT INTO ttl_table (key, col1) VALUES (2, 2) USING TTL {long_ttl};
        """)
    
        assert_all(
            self.session1,
            "SELECT * FROM ttl_table;",
            [[1, 1, None, None], [2, 2, None, None]]
        )
        time.sleep(short_ttl + 2)
        logger.debug('SStable dump after writes with RF = 1')
        print_sstable(self.node1, 'ks', 'ttl_table')
        print_sstable(self.node2, 'ks', 'ttl_table')
        self.session1 = self.patient_exclusive_cql_connection(self.node1)
        self.session1.execute("USE ks;")
        change_schema_safely(self.session1, nodes, """
            ALTER KEYSPACE ks WITH REPLICATION =
            {'class' : 'SimpleStrategy', 'replication_factor' : 2};
        """)
        self.cluster.nodetool('repair ks')
        logger.debug('SStable dump after alter RF to 2 and repair')
        print_sstable(self.node1, 'ks', 'ttl_table')
        print_sstable(self.node2, 'ks', 'ttl_table')
    
        self.node2.stop()
        ttl_start = time.time()
>       ttl_session1 = self.session1.execute('SELECT ttl(col1) FROM ttl_table;')

ttl_test.py:787: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
cassandra/cluster.py:2611: in cassandra.cluster.Session.execute
    ???
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

>   ???
E   cassandra.Unavailable: Error from server: code=1000 [Unavailable exception] message="Cannot achieve consistency level for cl ONE. Requires 1, alive 0" info={'consistency': 'ONE', 'required_replicas': 1, 'alive_replicas': 0}

cassandra/cluster.py:4829: Unavailable
```

**Comment by bhalevy** [comment]
https://jenkins.scylladb.com/job/scylla-enterprise/job/next/2320/artifact/dtest/logs-full.release.000/dtest-gw3.log
```
16:30:40,376 778     conftest                       INFO     conftest.py         :232  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Starting execution of test_ttl_is_respected_on_repair[parallel_aggregates_enabled] at 2023-04-04 16:30:40.376088
16:30:40,376 778     dtest_setup                    INFO     dtest_setup.py      :747  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: cluster ccm directory: /jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n
16:30:40,379 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Started cluster 'test' version 2023.2.0-dev installed in /jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.ccm/scylla-repository/b3321805ed554ecc61edb206d39d9825e1828ef0
16:30:40,379 778     dtest_setup                    DEBUG    dtest_setup.py      :670  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Scylla mode is 'release'
16:30:40,379 778     dtest_setup                    DEBUG    dtest_setup.py      :671  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Cluster *_request_timeout_in_ms=10000, range_request_timeout_in_ms=30000, cql request_timeout=30
16:30:40,386 778     dtest_setup                    DEBUG    dtest_setup.py      :715  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Done setting configuration options:
{   'consistent_cluster_management': False,
    'initial_token': None,
    'num_tokens': 256,
    'phi_convict_threshold': 5,
    'range_request_timeout_in_ms': 30000,
    'read_request_timeout_in_ms': 10000,
    'request_timeout_in_ms': 10000,
    'ring_delay_ms': 10000,
    'truncate_request_timeout_in_ms': 30000,
    'write_request_timeout_in_ms': 10000}
16:30:40,390 778     dtest_setup                    DEBUG    dtest_setup.py      :58   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Allocated cluster ID 2: /jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n
16:30:40,499 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: start_nodes: no_wait=False wait_for_binary_proto=True wait_other_notice=True force_wait_for_cluster_start=True
16:30:40,503 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: Starting scylla: args=['/jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n/test/node1/bin/scylla', '--options-file', '/jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n/test/node1/conf/scylla.yaml', '--log-to-stdout', '1', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--abort-on-internal-error', '1', '--logger-log-level', 'forward_service=trace', '--enable-parallelized-aggregation', 'true', '--api-address', '127.0.2.1', '--collectd-hostname', 'd092db558018.node1', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--collectd', '0', '--overprovisioned', '--prometheus-address', '127.0.2.1', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
16:31:05,066 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: Starting scylla-jmx: args=['/jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n/test/node1/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.2.1', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.2.1', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.2.1', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n/test/node1/bin/scylla-jmx-1.0.jar']
16:31:05,428 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: Starting scylla: args=['/jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n/test/node2/bin/scylla', '--options-file', '/jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n/test/node2/conf/scylla.yaml', '--log-to-stdout', '1', '--abort-on-seastar-bad-alloc', '--abort-on-lsa-bad-alloc', '1', '--abort-on-internal-error', '1', '--logger-log-level', 'forward_service=trace', '--enable-parallelized-aggregation', 'true', '--api-address', '127.0.2.2', '--collectd-hostname', 'd092db558018.node2', '--smp', '2', '--memory', '1024M', '--developer-mode', 'true', '--default-log-level', 'info', '--collectd', '0', '--overprovisioned', '--prometheus-address', '127.0.2.2', '--unsafe-bypass-fsync', '1', '--kernel-page-cache', '1', '--commitlog-use-o-dsync', '0', '--max-networking-io-control-blocks', '1000'] wait_other_notice=True wait_for_binary_proto=True
16:31:50,262 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: Starting scylla-jmx: args=['/jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n/test/node2/bin/symlinks/scylla-jmx', '-Dapiaddress=127.0.2.2', '-Djavax.management.builder.initial=com.scylladb.jmx.utils.APIBuilder', '-Djava.rmi.server.hostname=127.0.2.2', '-Dcom.sun.management.jmxremote', '-Dcom.sun.management.jmxremote.host=127.0.2.2', '-Dcom.sun.management.jmxremote.port=7199', '-Dcom.sun.management.jmxremote.rmi.port=7199', '-Dcom.sun.management.jmxremote.local.only=false', '-Xmx256m', '-XX:+UseSerialGC', '-Dcom.sun.management.jmxremote.authenticate=false', '-Dcom.sun.management.jmxremote.ssl=false', '-jar', '/jenkins/workspace/scylla-enterprise/gating-dtest-release/scylla-enterprise/.dtest/dtest-g2cxhn4n/test/node2/bin/scylla-jmx-1.0.jar']
16:31:50,622 778     cassandra.policies             INFO     policies.py         :289  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Using datacenter 'datacenter1' for DCAwareRoundRobinPolicy (via host '127.0.2.1:9042'); if incorrect, please specify a local_dc to the constructor, or limit contact points to local cluster nodes
16:31:50,622 778     cassandra.cluster              INFO     dtest_setup.py      :411  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: New Cassandra host <Host: 127.0.2.2:9042 datacenter1> discovered
16:32:04,884 778     ttl_test                       DEBUG    ttl_test.py         :771  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: SStable dump after writes with RF = 1
16:32:04,885 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=flush wait=True timeout=None
16:32:05,752 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=cfstats wait=True timeout=60
16:32:10,032 778     ttl_test                       DEBUG    ttl_test.py         :631  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: [node1] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-04-04T16:31:52.879523Z", "ttl" : 10, "expires_at" : "2023-04-04T16:32:02Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
][
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-04-04T16:31:52.881655Z", "ttl" : 1000, "expires_at" : "2023-04-04T16:48:32Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
]
16:32:10,033 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=flush wait=True timeout=None
16:32:10,898 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=cfstats wait=True timeout=60
16:32:14,090 778     ttl_test                       DEBUG    ttl_test.py         :631  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: [node2] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-04-04T16:31:52.879523Z", "ttl" : 10, "expires_at" : "2023-04-04T16:32:02Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
]
16:32:14,094 778     cassandra.cluster              INFO     dtest_setup.py      :411  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: New Cassandra host <Host: 127.0.2.2:9042 datacenter1> discovered
16:32:14,915 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=repair ks wait=True timeout=None
16:32:17,235 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=repair ks wait=True timeout=None
16:32:19,550 778     ttl_test                       DEBUG    ttl_test.py         :781  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: SStable dump after alter RF to 2 and repair
16:32:19,551 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=flush wait=True timeout=None
16:32:20,352 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node1: nodetool cmd=cfstats wait=True timeout=60
16:32:24,279 778     ttl_test                       DEBUG    ttl_test.py         :631  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: [node1] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-04-04T16:31:52.879523Z", "ttl" : 10, "expires_at" : "2023-04-04T16:32:02Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
][
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-04-04T16:31:52.881655Z", "ttl" : 1000, "expires_at" : "2023-04-04T16:48:32Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
]
16:32:24,280 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=flush wait=True timeout=None
16:32:25,069 778     ccm                            DEBUG    cluster.py          :709  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: node2: nodetool cmd=cfstats wait=True timeout=60
16:32:29,018 778     ttl_test                       DEBUG    ttl_test.py         :631  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: [node2] SStable dump of ks.ttl_table: [
  {
    "partition" : {
      "key" : [ "1" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-04-04T16:31:52.879523Z", "ttl" : 10, "expires_at" : "2023-04-04T16:32:02Z", "expired" : true },
        "cells" : [
          { "name" : "col1", "value" : 1 }
        ]
      }
    ]
  }
][
  {
    "partition" : {
      "key" : [ "2" ],
      "position" : 0
    },
    "rows" : [
      {
        "type" : "row",
        "position" : 18,
        "liveness_info" : { "tstamp" : "2023-04-04T16:31:52.881655Z", "ttl" : 1000, "expires_at" : "2023-04-04T16:48:32Z", "expired" : false },
        "cells" : [
          { "name" : "col1", "value" : 2 }
        ]
      }
    ]
  }
]
16:32:29,025 778     cassandra.cluster              WARNING  thread.py           :58   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Host 127.0.2.2:9042 has been marked down
16:32:30,027 778     cassandra.pool                 WARNING  thread.py           :58   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Error attempting to reconnect to 127.0.2.2:9042, scheduling retry in 1.86 seconds: [Errno 111] Tried connecting to [('127.0.2.2', 9042)]. Last error: Connection refused
16:32:31,929 778     cassandra.pool                 WARNING  thread.py           :58   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Error attempting to reconnect to 127.0.2.2:9042, scheduling retry in 3.88 seconds: [Errno 111] Tried connecting to [('127.0.2.2', 9042)]. Last error: Connection refused
16:32:32,021 778     dtest_class                    DEBUG    dtest_class.py      :44   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Retrying request after UE. Attempt #0
16:32:32,021 778     dtest_class                    DEBUG    dtest_class.py      :44   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Retrying request after UE. Attempt #1
16:32:32,022 778     dtest_class                    DEBUG    dtest_class.py      :44   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Retrying request after UE. Attempt #2
16:32:32,022 778     dtest_class                    DEBUG    dtest_class.py      :44   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Retrying request after UE. Attempt #3
16:32:32,023 778     dtest_class                    DEBUG    dtest_class.py      :44   | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: Retrying request after UE. Attempt #4
16:32:32,107 778     errors                         ERROR    conftest.py         :203  | test_ttl_is_respected_on_repair[parallel_aggregates_enabled]: test failed: 
```

https://jenkins.scylladb.com/job/scylla-enterprise/job/next/2320/artifact/dtest/logs-full.release.000/1680625955141_ttl_test.py%3A%3ATestDistributedTTL%3A%3Atest_ttl_is_respected_on_repair%5Bparallel_aggregates_enabled%5D/node1.log
```
INFO  2023-04-04 16:32:29,024 [shard 0] gossip - Got shutdown message from 127.0.2.2, received_generation=1680625865, local_generation=1680625865
INFO  2023-04-04 16:32:29,025 [shard 0] gossip - InetAddress 127.0.2.2 is now DOWN, status = shutdown
```

It looks like the test needs to stop node2 with `wait_other_notice=True`
Maybe it should even be the default, while only tests that want to stress the system can call stop(gently=False, wait_other_notice=False)

**Comment by bhalevy** [comment]
@fruch any idea why the retries above are so tight?
It looks like the driver doesn't wait between retries at all (or less than a millisecond)

**Comment by bhalevy** [comment]
cc @Jadw1

**Comment by fruch** [comment]
> @fruch any idea why the retries above are so tight?
> It looks like the driver doesn't wait between retries at all (or less than a millisecond)

The driver is quite dumb in the area of retries, the user doesn't get any control for specifying the time between retries.

Which prevents easily implementing exponential backoffs.

We opened an issue a few years back, but no one got implementing it

(And even if we did, we still need to switch to our fork, or support both)

**Comment by bhalevy** [comment]
> > @fruch any idea why the retries above are so tight?
> > It looks like the driver doesn't wait between retries at all (or less than a millisecond)
> 
> The driver is quite dumb in the area of retries, the user doesn't get any control for specifying the time between retries.
> 
> Which prevents easily implementing exponential backoffs.
> 
> We opened an issue a few years back, but no one got implementing it
> 
> (And even if we did, we still need to switch to our fork, or support both)

So maybe the handler should sleep for now.
Or keep the default handler for unavailable without overriding it:
```
        return (self.RETRY_NEXT_HOST, None) if retry_num == 0 else (self.RETHROW, None)
```

Or just return RETHROW instead of causing a burst of retries.

**Comment by fruch** [comment]
It might help, but I think it keeps all the requests from taking place until the sleep ends.

**Comment by bhalevy** [comment]
https://jenkins.scylladb.com/job/scylla-master/job/next/5859/testReport/junit/ttl_test/TestDistributedTTL/Tests___Sanity_Tests___test_ttl_is_respected_on_repair_parallel_aggregates_disabled_/
https://jenkins.scylladb.com/job/scylla-master/job/next/5859/testReport/junit/ttl_test/TestDistributedTTL/Tests___Sanity_Tests___test_ttl_is_respected_on_repair_parallel_aggregates_enabled_/
```AssertionError: Expected [[1, 1, None, None], [2, 2, None, None]] from SELECT * FROM ttl_table;, but got [[1, 1, None, None]]
```

**Comment by DoronArazii** [comment]
@Jadw1 any plans to deal with it soon?

**Comment by Jadw1** [comment]
> @Jadw1 any plans to deal with it soon?

I don't have any idea what to do with the test :/

**Comment by bhalevy** [comment]
> > @Jadw1 any plans to deal with it soon?
> 
> I don't have any idea what to do with the test :/

I'm not sure the problem is with the test at this point.
We need to reproduce and understand what's going on on the server side before determining what should be fixed next.

**Comment by Jadw1** [comment]
> I'm not sure the problem is with the test at this point. We need to reproduce and understand what's going on on the server side before determining what should be fixed next.
I've been trying to reproduce any of ttl_test errors locally lots of times but without success

**Comment by DoronArazii** [comment]
@Jadw1 / @nyh any progress here ?

it's a p1 issue that should have attention

**Comment by nyh** [comment]
@DoronArazii I was never involved in this issue so I don't have any proress but I'll take a look now.

Frankly I don't understand why after @bhalevy 's comment nobody from QA is involved to think if it's a Scylla bug or a test bug, and whether this test should be removed from gating or not.

**Comment by nyh** [comment]
> Frankly I don't understand why after @bhalevy 's comment nobody from QA is involved to think if it's a Scylla bug or a test bug, and whether this test should be removed from gating or not.

I see it was already removed from gating:
```
    # Removed temporarily from gating de to #13137 / #13276
    # @pytest.mark.next_gating
    @pytest.mark.dtest_debug
    @pytest.mark.parametrize("enable_parallized_aggregation", ['true', 'false'], ids=["parallel_aggregates_enabled", 
"parallel_aggregates_disabled"])
    def test_ttl_is_respected_on_repair(self, enable_parallized_aggregation):
```

So why is this p1?
```

**Comment by nyh** [comment]
Unless someone can give me proof otherwise, my conclusion this is an **outdated** test bug that was already fixed by @bhalevy 's https://github.com/scylladb/scylla-dtest/commit/ff14d9329c805e298cc5c12726a6904f887e3ee1.

I think the test bug was that we changed the RF from 1 to 2 and didn't wait enough for this to be recognized, so when the test does 
```python
        self.node2.stop()
        ttl_session1 = self.session1.execute('SELECT ttl(col1) FROM ttl_table;').one()[0]
```

node1 receives the read request and is convinced that only node2 can answer it (because it wrongly thinks RF=1), but node2 is down.

As far as I can tell enterprise 2023.1 branch didn't get @bhalevy 's fix (Benny?)  which is why the bug returned on that branch two weeks ago, but hasn't been seen again on master. @bhalevy or @fruch please backport the test fix if.

**Comment by nyh** [comment]
@fruch did you backport them now or before the last failure?
Also as I said, it was removed from gating on master so I'm not surprised it didn't fail as gating. But the last failure that @DoronArazii brought me in for now (?) was in the 2023.1 branch, I don't know if it was removed from gating there or not.

In any case, looks like a test bug or missing backport so I don't have much more to contribute to this issue unless someone is convinced the issue still exists on master dtest branch. If it doesn't, let's close this issue instead of alarming people that it's p1...

**Comment by bhalevy** [comment]
As I explained in https://github.com/scylladb/scylla-dtest/pull/3111, the test was removed from gating since it kept failing in next promotion despite the test fixes: https://github.com/scylladb/scylla-dtest/pull/3086, https://github.com/scylladb/scylla-dtest/pull/3102.

For some reason it tended to fail more in next gating than in dtest release. Maybe we use different hardware that has different timing there.

We still need to determine what doesn't work between the Python client `wait_for_schema_agreement` and scylla.

**Comment by mykaul** [comment]
> Lately, the number of CI-Stability cases was increased - and I was sent (by people with ranks) to interrupt developers with their ongoing tasks, so we all enjoy from a stabilized CI.

Paraphrasing a bit, CI stability is one of the goals of the developers, handled by the developers, for the developers - ranks are unrelated.

**Comment by Jadw1** [comment]
I have an idea why the problem might have occured.

When the issue was reported, `assert_row_count` was executed just after stopping `node1`:
https://github.com/scylladb/scylla-dtest/blob/060132cb68be3c6248e05f5c56e6881817f14fa7/ttl_test.py#L789-L791
```
self.node1.stop()

assert_row_count(session2, 'ttl_table', 1)
```

The error says: `Cannot achieve consistency level for cl ONE. Requires 1, alive 0` and the error happened in `parallel_aggregates_enabled` mode, so my guess is that `SELECT COUNT(*)` was executed before `node2` found out that `node1` is down, so it tried to parallelize the aggregation query and the error happened.
I don't know exactly how getting alive replica works(it is used in aggregate parallelization) or is `node.stop()` from python driver a blocking operation, but it looks like possible situation.

On the other hand, in `forward_service`, which 
is responsible for aggregates distribution, there is a `retrying_dispatcher` that should retried the failed sub-query on the super node(e.g. in case the destination node is down). So this is my only concern that sth is not ok.

However, there was a commit https://github.com/scylladb/scylla-dtest/commit/e369279e8415ae2d9c3a8d032ade116d1e878314 that added some instructions between those 2 lines mentioned above: https://github.com/scylladb/scylla-dtest/blob/e369279e8415ae2d9c3a8d032ade116d1e878314/ttl_test.py#L786-L787
@fruch @bhalevy does this error is still happening from time to time, or did it disappear after this commit?

Summing up, my proposition is I will create a test for `retrying_dispatcher` and test the mechanism of retrying subqueries. If the tests are ok, we can consider this issue is resolved and close it. Meanwhile, we can leave logging level of `forward_service` on `trace` and we can inspect the nodes' logs if the error appears ever again.

Cc @mykaul @eliransin @ptrsmrn @DoronArazii

**Comment by kbr-scylla** [comment]
> Wouldn't it be enough to pause the node? If yes, then you can add this ability to test.py (pylib) or dtest (ccm) and write the test there.
> Pausing should be easy to add — you can pause a process by sending SIGSTOP to it and unpause by sending SIGCONT.

https://github.com/scylladb/scylladb/pull/13850/commits/97484972559adcaac92d5311fee2bfc29fc060f4

**Comment by Jadw1** [comment]
https://github.com/scylladb/scylla-dtest/pull/3234 - tests of retrying mechanism
https://github.com/scylladb/scylla-dtest/pull/3235 - wait_for_others

But I still couldn't reproduce the error. When RF=2 and one of the node was stopped, retrying mechanism worked correctly and the query returned correct result. I played with some sleeps but it also didn't help.

If the RF=1, the query obviously failed but with errors (depends if node discovered the other node is down)
```
OperationTimedOut: errors={'127.0.0.1:9042': 'Client request timeout. See Session.execute[_async](timeout)'}, last_host=127.0.0.1:9042
```
or 
```
NoHostAvailable: ('Unable to complete the operation against any hosts', {<Host: 127.0.0.1:9042 datacenter1>: Unavailable('Error from server: code=1000 [Unavailable exception] message="Cannot achieve consistency level for cl ONE. Requires 1, alive 0" info={\'consistency\': \'ONE\', \'required_replicas\': 1, \'alive_replicas\': 0}')})
```

**Comment by eliransin** [comment]
Both myself and @Jadw1 believe that following the changes to the test and scylla core done by @michoecho and @bhalevy 
fixed this and that it never had anything to do with the parallel aggregates.
We will wait for another week and if this failure doesn't happen again we will close the issue.

**Comment by bhalevy** [comment]
It actually happened yesterday but https://github.com/scylladb/scylladb/pull/14158 is not merged yet, so it's not unexpected...

**Comment by michoecho** [comment]
I think https://github.com/scylladb/scylladb/issues/13137 does explain also this issue. The test expects RF=2, but one shard is stuck thinking that RF=1, so when it gets a query to a range owned by the other node (which is stopped), it responds with `"Cannot achieve consistency level for cl ONE. Requires 1, alive 0"`.

Retrying can't help because the node doesn't have anywhere to retry to. It doesn't know that it's also supposed to replicate the range.

**Comment by eliransin** [comment]
Thanks @bhalevy @michoecho , I will verify that this doesn't happen for a while after #14158 is merged and only then
will close.
I realize that most chances are that @michoecho fix is going to make this go away but we had a lot of attempts before to fix this test.
I will, however, reduce it's priority :slightly_smiling_face:

**Comment by bhalevy** [comment]
@eliransin we should be able to close it now with #14158

**Comment by tchaikov** [comment]
i am closing it.
